accepted/2024/observable_collection.md

```diff
@@ -0,0 +1,201 @@
+  - Pro: Core API is simple while providing an affordance to the user to solve
+    the problem without requiring changes from the UI framework.
+  - Con: More complex state, might not play well with WPF's `CollectionView`.
+  - > [!NOTE]
+    > Doesn't seem feasible. Handlers of the event likely assume the collection
+    > was just changed to include the one modification. Just translating the
+    > event would mean they get a series of events but the handler sees the
+    > final collection in all invocations.
+
+* **Event enumerator**. We could expose a method on
+  `NotifyCollectionChangedEventArgs` that returns an
+  `IEnumerable<NotifyCollectionChangedEventArgs>`. For single item events it
+  returns itself, otherwise it creates a new event arg per item. This requires
+  UI controls to make code changes if they don't support bulk notifications, but
+  the change is fairly targeted.
+  - Pro: Simple API
+  - Con: Still breaks consumers, just provides an easier path to adapt to the
+    new behavior without having to fully support bulk notifications.
+  - > [!NOTE]
+    > Doesn't seem feasible. Handlers of the event likely assume the collection
+    > was just changed to include the one modification. Just translating the
+    > event would mean they get a series of events but the handler sees the
+    > final collection in all invocations.
+
+* **Handling it at the CollectionView level**. WPF seems to have an indirection
+  between the data-bound UI controls and the observable collections via the
+  collection view. Maybe we could expose a property on `CollectionView` that
+  controls whether the collection view will translate bulk events to individual
+  events.
+  - Pro: Simpler API for the base, with an opt-out switch at the level of WPF.
+  - Con: Needs more investigation to determine feasibility, would only be helpful for WPF
+
+* **Separate API**. We could add a new interface
+  `IBatchedNotifyCollectionChanged` interface that `ObservableCollection<T>`
+  also implements. The interface adds new even `BulkCollectionChanged`.
+  - Pro: Consumers opt-into bulk events by subscribing to a different event.
+    This allows a single instance of `ObservableCollection<T>` to be bound to
+    multiple UI elements and leverage bulk updates when it's supported while
+    falling back when it's not. This would also allow change of semantics to
+    allow for non-contiguous updates, for instance.
+  - Con: Complicates that API; requires all consumers to explicitly opt-in, even
+    if happen to already support bulk notifications. Would also likely require
+    adding a corresponding interface to WinUI that .NET can project to. Will
+    likely take longer to be supported by UI frameworks.
+  - > [!NOTE]
+    > For the same reason event adapters aren't viable just having a new
+    > interface is probably not sufficient. Rather, we'd need a different
+    > implementation like `BulkObservableCollection<T>`. However, for WinUI
```

Comment:
We looked into this before and we concluded that actually we would not need a new interface at all for both WinUI 3 and UWP. `INotifyCollectionChanged` is projected to either `Microsoft.UI.Xaml.Interop.INotifyCollectionChanged` (for WinUI 3) or `Windows.UI.Xaml.Interop.INotifyCollectionChanged` (for UWP), and these interfaces already have all the support they need for multiple items (they kinda mirror the .NET event args). Which also means the underlying infrastructure (ie. WinUI 3 and UWP XAML controls) already work in term of that, we'd just need to do additional testing to be extra sure. But in general, all of this should "just work" for both WinUI 3 and modern UWP.

cc. @manodasanW

Reply:
Yeah, I think if any work would be within like `ListView` vs. the projections of the data, similar to the work needed in WPF.

Reply:
That's cool!

Presumably, the assumption still is that the ranges are contiguous, correct?

Reply:
Looking at the properties for [`NotifyCollectionChangedEventArgs`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.specialized.notifycollectionchangedeventargs) I assume so, yes. @jkoritzinsky do you perhaps know? Otherwise if we need we can reach out to XAML folks to confirm (anyone from the WinUI team), surely can't hurt.

Reply:
Yes, the two event args types have a 1:1 projection; that is each can represent exactly the same set of states as the other. So if the .NET version only supports contiguous ranges, the WinRT versions also support only contiguous ranges.

Reply:
Adding @ranjeshj who can maybe comment more on the UI work beyond the projection capabilities for the containerization aspects.